modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+    moved around from concurrent additions or removals may be missed while clearing.
+  */
+  proc clear() {
+    var localThis = getPrivatizedThis;
+    coforall loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      forall segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        if segment.acquireIfNonEmpty(STATUS_REMOVE) {
+          var block = segment.headBlock;
+          while block != nil {
+            var tmp = block;
+            block = block.next;
+            delete tmp;
+          }
+
+          segment.headBlock = nil;
+          segment.tailBlock = nil;
+          segment.nElems.write(0);
+          segment.releaseStatus();
+        }
+      }
+    }
+  }
+
+  /*
+    Triggers a more static approach to load balancing, fairly redistributing all
+    elements fairly for bags across nodes. The result will result in all segments
+    having roughly the same amount of elements.
+
+    **Note:** This method is very heavy-weight in that it should not be called too
+    often. Dynamic work stealing handles cases where there is a relatively fair
+    distribution across majority of nodes, but this should be called when you have
+    a severe imbalance, or when you have a smaller number of elements to balance.
+    Furthermore, while this operation is parallel-safe, it should be called in a
+    sequential context.
+  */
+  proc balance() {
+    var localThis = getPrivatizedThis;
+    // Phase 1: Acquire all locks from from first node and segment to last
+    // node and segment (our global locking order...)
+    for loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      for segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        segment.acquire(STATUS_BALANCE);
+      }
+    }
+
+    // Phase 2: Concurrently redistribute elements from segments which contain
+    // more than the computed average.
+    coforall segmentIdx in 0 .. #here.maxTaskPar {
+      var nSegmentElems : [localThis.targetLocales.size] int;
+      var locIdx = 0;
+      for loc in localThis.targetLocales do on loc {
+        var nElems = getPrivatizedThis.bag.segments[segmentIdx].nElems.read() : int;
+        nSegmentElems[locIdx] = nElems;
+        locIdx += 1;
+      }
+
+      // Find the average and the excess. The excess is calculated as the amount
+      // of elements a segment has over the average, which is used to calculate
+      // the buffer size for each segment.
+      var total = (+ reduce nSegmentElems);
+      var avg = total / locIdx;
+      var excess : int;
+      for nElems in nSegmentElems {
+        if nElems > avg {
+          excess += nElems - avg;
+        }
+      }
+
+      // Allocate buffer, which holds the 'excess' elements for redistribution.
+      // Then fill it.
+      var buffer = c_malloc(eltType, excess);

```

Comment:
Currently I have a helper method that takes both a c_ptr, a size, and a locale and writes into it, and one that does the opposite (reads from a c_ptr). I chose c_ptr over chapel arrays as they are more lightweight, but I didn't know about the bulk transfer support you mentioned before.